FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1139,377 +1116,688 @@ namespace edm {
+                          handleNextItemAfterMergingRunEntries(status, holder);
+
+                          CMS_SA_ALLOW try {
+                            streamQueuesInserter_.push(
+                                *holder.group(), [this, status, precedingTasksSucceeded, holder]() mutable {
+                                  for (unsigned int i = 0; i < preallocations_.numberOfStreams(); ++i) {
+                                    CMS_SA_ALLOW try {
+                                      streamQueues_[i].push(
+                                          *holder.group(),
+                                          [this, i, status, precedingTasksSucceeded, holder]() mutable {
+                                            streamBeginRunAsync(
+                                                i, std::move(status), precedingTasksSucceeded, std::move(holder));
+                                          });
+                                    } catch (...) {
+                                      if (status->streamFinishedBeginRun()) {
+                                        WaitingTaskHolder copyHolder(holder);
+                                        copyHolder.doneWaiting(std::current_exception());
+                                        status->resetBeginResources();
+                                        queueWhichWaitsForIOVsToFinish_.resume();
+                                        exceptionRunStatus_ = status;
+                                      }
+                                    }
+                                  }
+                                });
+                          } catch (...) {
+                            WaitingTaskHolder copyHolder(holder);
+                            copyHolder.doneWaiting(std::current_exception());
+                            status->resetBeginResources();
+                            queueWhichWaitsForIOVsToFinish_.resume();
+                            exceptionRunStatus_ = status;
+                          }
+                        }) | runLast(postSourceTask);
+                      } catch (...) {
+                        status->resetBeginResources();
+                        queueWhichWaitsForIOVsToFinish_.resume();
+                        status->resumeGlobalRunQueue();
+                        postSourceTask.doneWaiting(std::current_exception());
+                      }
+                    });  // task in sourceResourcesAcquirer
+              } catch (...) {
+                status->resetBeginResources();
+                queueWhichWaitsForIOVsToFinish_.resume();
+                status->resumeGlobalRunQueue();
+                postRunQueueTask.doneWaiting(std::current_exception());
+              }
+            });  // task in runQueue
+      } catch (...) {
+        status->resetBeginResources();
+        queueWhichWaitsForIOVsToFinish_.resume();
+        nextTask.doneWaiting(std::current_exception());
       }
+    }) | chain::runLast(std::move(iHolder));
```

Comment:
I followed the same style as beginLumiAsync (~150 lines), but beginRunAsync is even bigger (~200 lines). I had the same thoughts and was thinking about this. If you want I will do it now. It does not look hard. I could split both of them up. The most obvious approach would be to put the contents of each large lambda in a member function of EventProcessor and the lambda just calls the function. There might be a better way.

I don't think there is any advantage for the purpose of code reuse. Possibly it would make the code easier to read and understand. Maybe... It would certainly make code reviewers in general more happy. General advice is to have smaller functions.

Ask if you want me to try it now. Otherwise, I will just leave this for a later PR.